crates/game/pong/src/game.rs

```diff
@@ -0,0 +1,166 @@
+use game_lib::Game;
+use ggez::glam::{vec2, Vec2};
+use rand::random_range;
+use std::f32::consts::{FRAC_PI_4, PI};
+use std::time::Duration;
+
+pub const MAX_BOUNCE_ANGLE: f32 = FRAC_PI_4;
+pub const PLAYER_HEIGHT: f32 = 0.2;
+pub const PLAYER_WIDTH: f32 = 0.02;
+
+pub struct PongGame {
+    pub player: (PongPlayer, PongPlayer),
+    pub state: PongGameState,
+}
+
+pub struct PongGameState {
+    pub score: (usize, usize),
+    pub ball_pos: Vec2,
+    pub ball_dir: Vec2,
+}
+
+impl PongGame {
+    pub fn new(player_one: PongPlayer, player_two: PongPlayer) -> Self {
+        let start_direction = random_range(0.0..(PI * 2.));
+
+        PongGame {
+            player: (player_one, player_two),
+            state: PongGameState {
+                score: (0, 0),
+                ball_pos: vec2(0.5, 0.5),
+                ball_dir: Vec2::from_angle(start_direction),
+            },
+        }
+    }
+
+    fn reset_ball(&mut self) {
+        let start_direction = random_range(0.0..(PI * 2.));
+
+        self.state.ball_pos = vec2(0.5, 0.5);
+        self.state.ball_dir = Vec2::from_angle(start_direction);
+    }
+}
+
+impl Game for PongGame {
+    fn tick(&mut self, delta_time: Duration) {
+        // update positions
+        self.player.0.update_pos(&self.state, &delta_time);
+        self.player.1.update_pos(&self.state, &delta_time);
+
+        self.state.ball_pos += self.state.ball_dir * delta_time.as_secs_f32();
+
+        // bounce ball on ceiling / floor
+        if self.state.ball_pos.y > 1. {
+            self.state.ball_dir.y = -self.state.ball_dir.y.abs();
+        }
+
+        if self.state.ball_pos.y < 0. {
+            self.state.ball_dir.y = self.state.ball_dir.y.abs();
+        }
+
+        // bounce ball on panels
+        if self.state.ball_pos.x > 1. - PLAYER_WIDTH
+            && self.player.1.does_intersect_ball(&self.state.ball_pos)
+        {
+            let bounce_angle = self.player.1.get_bounce_angle(self.state.ball_pos.y);
+
+            self.state.ball_dir.x = -self.state.ball_dir.x.abs() * bounce_angle.cos();
+            self.state.ball_dir.y = bounce_angle.sin();
+        }
+
+        if self.state.ball_pos.x < PLAYER_WIDTH
+            && self.player.0.does_intersect_ball(&self.state.ball_pos)
+        {
+            let bounce_angle = self.player.0.get_bounce_angle(self.state.ball_pos.y);
+
+            self.state.ball_dir.x = self.state.ball_dir.x.abs() * bounce_angle.cos();
+            self.state.ball_dir.y = bounce_angle.sin();
+        }
+
+        // handle player loose
+        if self.state.ball_pos.x > 1. {
+            self.state.score.0 += 1;
+            self.reset_ball()
+        }
+
+        if self.state.ball_pos.x < 0. {
+            self.state.score.1 += 1;
+            self.reset_ball()
+        }
```

Comment:
Game trait should offer functions to stop the game. Right?

Reply:
we can...
but is it need?
like what would be the usage for the model trainer?

Reply:
I thought of the trainer that eg says the model to make a point in pong. The model then tries. If it looses, or takes long it gets an panelty. If it wins it gets a reward. This result is then returned to the trainer. The trainer updates the model and starts another round.

Reply:
I would just say create a new game instead of stopping and restarting the old one